RFC-0021-Distributed-Pipeline-Parallel-Technical.md

```diff
@@ -0,0 +1,651 @@
+* (**D5**) Does not require model to be an `nn.Sequential`
+* (**D8**) User’s original training loop is preserved with only slight modifications (`@smp.step` annotation and other things)
+
+**Con**
+
+* (**D2**) (split pro/con) Support for *some* kind of schedules, but not strictly implemented as those described in the literature
+* (**D6**) (not sure about this one) Seems to still rely on synchronous training loop. May need modifications to support async (but not sure if there are any fundamental limitations?)
+
+## Approach 6: SPMD with Program capture/JIT compilation and message passing (OneFlow)
+
+[OneFlow](https://arxiv.org/abs/2110.15032) is a deep learning framework that purports to redesign the deep learning programming model to better support distributed computation. The framework uses a “consistent view” abstraction to represent distributed memory. The framework sports a compiler that can run global optimization of device placement. It also uses a unified actor model abstraction for its distributed runtime.
+
+An example of using OneFlow for pipeline parallelism can be seen in this [tutorial](https://docs.oneflow.org/en/master/parallelism/06_pipeline.html). The example instantiates a sequential two-stage model architecture and uses the `to_consistent` API to move the submodules to the appropriate CUDA devices. It also sets `config.stage_id` on each of the submodules to give a monotonically increasing stage number. Finally, on the `nn.Graph` it uses `config.set_gradient_accumulation_steps` to delay optimization for 2 micro-batches, and calls `add_optimizer` to add the optimizer class.
+
+`oneflow.distributed.launch` will launch the processes for each “actor”. Eager mode and graph mode reportedly go down the same (or similar) code path for distributed graph processing, with eager mode using a LazyTensor-like model to capture the program. The compiler knows which rank the compilation is for and will emit code specifically for that rank. Micro-batch splitting and `1f1b` seem to be hard-coded, if they are implemented.
+
+### Pros and Cons of the Approach
+
+**Pro**
+
+* (split pro/con) (**D2**) Not clear if `1f1b` or other schedules are implemented
+* (**D3/D4**) Composable with other parallelism schemes via their “Consistent View” definition
+* (**D5**) `nn.Sequential` not needed, but potentially an `nn.Graph` instance may be needed in some cases
+* (**D6**) Async is probably supportable but not clear. From their presentation, the actor/register model with backpressure can implement on-demand data loading, but I’m not 100% sure what that API looks like
+* (**D7**) Unified programming model that already exists
+
+**Con**
+
+* (split pro/con) (**D2**) Not clear if `1f1b` or other schedules are implemented/implementable?
+* (**D8**) Not clear what the training loop abstraction looks like. The optimizer is installed via an `nn.Graph` API. Loss calculation is created in the `nn.Graph.build()` method.
+
+## Approach 7: Varuna
```

Comment:
@stas00 I've added details about Varuna in this section, please have a look and let me know if I've missed anything